File: packages/sdk-js/src/core/negotiation.ts
import { ContractClient } from "../contracts/client";
import { StrategyEngine } from "./strategy";
import { NEGOTIATION_DEFAULTS } from "../config";

export class NegotiationSkill {
  private client: ContractClient;
  private strategy: StrategyEngine;
  public maxRounds: number;
  public pollInterval: number;

  constructor(client: ContractClient, strategy: StrategyEngine, maxRounds?: number, pollInterval?: number) {
    this.client = client;
    this.strategy = strategy;
    this.maxRounds = maxRounds || NEGOTIATION_DEFAULTS.maxRounds;
    this.pollInterval = pollInterval || NEGOTIATION_DEFAULTS.pollIntervalMs;
  }

  static usdcToWei(usdc: number): bigint {
    return BigInt(Math.floor((usdc / 1000) * 1e18));
  }

  static weiToUsdc(wei: number): number {
    return (wei / 1e18) * 1000;
  }

  async openSession(seller: string, description: string, budgetCapUsdc: number, durationSeconds = 600): Promise<string> {
    const budgetWei = NegotiationSkill.usdcToWei(budgetCapUsdc);
    return this.client.openSession(seller, description, budgetWei, this.maxRounds, durationSeconds);
  }

  async runBuyerLoop(
    sessionId: string, budgetCapUsdc: number, sellerReputation: number,
    serviceDescription: string, onDeal: (session: any) => void
  ): Promise<any> {
    let lastSubmittedRound = -1;

    while (true) {
      await this.sleep(this.pollInterval);
      const session = await this.client.getSession(sessionId);

      if (session.status === "AGREED") {
        console.log(`[Negotiation] Deal at ${NegotiationSkill.weiToUsdc(session.agreedPrice).toFixed(4)} USDC`);
        onDeal(session);
        return session;
      }

      if (session.status === "EXPIRED" || session.status === "CANCELLED") {
        console.log(`[Negotiation] Session ended: ${session.status}`);
        return session;
      }

      const buyerOffers = session.buyerOffers;
      const sellerOffers = session.sellerOffers;
      const currentRound = session.currentRound;

      if (sellerOffers.length > buyerOffers.length && currentRound === buyerOffers.length && currentRound !== lastSubmittedRound) {
        const latestAsk = NegotiationSkill.weiToUsdc(sellerOffers[sellerOffers.length - 1]);
        const bidsUsdc = buyerOffers.map(NegotiationSkill.weiToUsdc);
        const asksUsdc = sellerOffers.map(NegotiationSkill.weiToUsdc);

        const result = await this.strategy.buyerStrategy({
          currentAskUsdc: latestAsk,
          budgetCapUsdc,
          currentRound,
          maxRounds: this.maxRounds,
          sellerReputation,
          bidHistory: bidsUsdc,
          askHistory: asksUsdc,
          serviceDescription,
        });

        const bidWei = NegotiationSkill.usdcToWei(result.offer);
        await this.client.submitBuyerBid(sessionId, bidWei);
        lastSubmittedRound = currentRound;
        console.log(`[Negotiation] Round ${currentRound + 1} bid: ${result.offer.toFixed(4)} USDC — ${result.reasoning}`);
      }
    }
  }

  async runSellerLoop(
    sellerAddress: string, floorPriceUsdc: number, serviceDescription: string, onDeal: (session: any) => void
  ): Promise<void> {
    const handledSessions = new Set<string>();
    const lastSubmitted = new Map<string, boolean>();

    while (true) {
      await this.sleep(this.pollInterval);

      let openSessions: string[];
      try {
        openSessions = await this.client.getOpenSessionsForSeller(sellerAddress);
      } catch {
        continue;
      }

      for (const sessionId of openSessions) {
        if (handledSessions.has(sessionId)) continue;
        const session = await this.client.getSession(sessionId);
        const buyerRep = await this.client.getReputationDisplay(session.buyer);
        if (buyerRep < 1.0) {
          console.log(`[Negotiation] Ignoring low-rep buyer: ${buyerRep.toFixed(1)}/10`);
          handledSessions.add(sessionId);
          continue;
        }
        console.log(`[Negotiation] New session from buyer rep ${buyerRep.toFixed(1)}/10`);
        handledSessions.add(sessionId);
      }

      for (const sessionHex of handledSessions) {
        const session = await this.client.getSession(sessionHex);

        if (session.status === "AGREED") {
          onDeal(session);
          continue;
        }
        if (session.status === "EXPIRED" || session.status === "CANCELLED") continue;

        const buyerOffers = session.buyerOffers;
        const sellerOffers = session.sellerOffers;
        const currentRound = session.currentRound;
        const roundKey = `${sessionHex}:${currentRound}`;

        if (buyerOffers.length > sellerOffers.length && currentRound === sellerOffers.length && !lastSubmitted.has(roundKey)) {
          const latestBid = buyerOffers.length > 0 ? NegotiationSkill.weiToUsdc(buyerOffers[buyerOffers.length - 1]) : 0;
          const bidsUsdc = buyerOffers.map(NegotiationSkill.weiToUsdc);
          const asksUsdc = sellerOffers.map(NegotiationSkill.weiToUsdc);
          const buyerRep = await this.client.getReputationDisplay(session.buyer);

          if (buyerOffers.length === 0) {
            const initialAsk = floorPriceUsdc * 2.5;
            const askWei = NegotiationSkill.usdcToWei(initialAsk);
            await this.client.submitSellerAsk(sessionHex, askWei);
            lastSubmitted.set(roundKey, true);
            console.log(`[Negotiation] Initial ask: ${initialAsk.toFixed(4)} USDC`);
            continue;
          }

          const result = await this.strategy.sellerStrategy({
            currentBidUsdc: latestBid,
            floorPriceUsdc,
            currentRound,
            maxRounds: this.maxRounds,
            buyerReputation: buyerRep * 1000,
            askHistory: asksUsdc,
            bidHistory: bidsUsdc,
            serviceDescription,
          });

          const askWei = NegotiationSkill.usdcToWei(result.offer);
          await this.client.submitSellerAsk(sessionHex, askWei);
          lastSubmitted.set(roundKey, true);
          console.log(`[Negotiation] Round ${currentRound + 1} ask: ${result.offer.toFixed(4)} USDC — ${result.reasoning}`);
        }
      }
    }
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: packages/sdk-python/anp_sdk/core/negotiation.py
import time
from web3 import Web3
from ..contracts.client import ContractClient
from .strategy import StrategyEngine
from ..config import NEGOTIATION_DEFAULTS


class NegotiationSkill:

    def __init__(self, client: ContractClient, strategy: StrategyEngine,
                 max_rounds: int = None, poll_interval: int = None):
        self.client = client
        self.strategy = strategy
        self.max_rounds = max_rounds or NEGOTIATION_DEFAULTS["max_rounds"]
        self.poll_interval = poll_interval or NEGOTIATION_DEFAULTS["poll_interval"]

    # ── Buyer side ─────────────────────────────────────────────────

    def open_session(self, seller: str, description: str,
                     budget_cap_usdc: float, duration_seconds: int = 600) -> bytes:
        budget_wei = self._usdc_to_wei(budget_cap_usdc)
        session_id = self.client.open_session(
            seller=seller, description=description, budget_cap_wei=budget_wei,
            max_rounds=self.max_rounds, duration_seconds=duration_seconds
        )
        print(f"[Negotiation] Session opened: {session_id.hex()[:16]}...")
        return session_id

    def run_buyer_loop(self, session_id: bytes, budget_cap_usdc: float,
                       seller_reputation: float, service_description: str,
                       on_deal: callable) -> dict | None:
        last_submitted_round = -1
        while True:
            time.sleep(self.poll_interval)
            session = self.client.get_session(session_id)

            if session["status"] == "AGREED":
                print(f"[Negotiation] Deal at {self._wei_to_usdc(session['agreedPrice']):.4f} USDC")
                on_deal(session)
                return session

            if session["status"] in ("EXPIRED", "CANCELLED"):
                print(f"[Negotiation] Session ended: {session['status']}")
                return session

            buyer_offers = session["buyerOffers"]
            seller_offers = session["sellerOffers"]
            current_round = session["currentRound"]

            if (len(seller_offers) > len(buyer_offers) and
                current_round == len(buyer_offers) and
                current_round != last_submitted_round):

                latest_ask = self._wei_to_usdc(seller_offers[-1])
                bids_usdc = [self._wei_to_usdc(b) for b in buyer_offers]
                asks_usdc = [self._wei_to_usdc(a) for a in seller_offers]

                result = self.strategy.buyer_strategy(
                    current_ask_usdc=latest_ask, budget_cap_usdc=budget_cap_usdc,
                    current_round=current_round, max_rounds=self.max_rounds,
                    seller_reputation=seller_reputation, bid_history=bids_usdc,
                    ask_history=asks_usdc, service_description=service_description,
                )

                bid_wei = self._usdc_to_wei(result["offer"])
                self.client.submit_buyer_bid(session_id, bid_wei)
                last_submitted_round = current_round
                print(f"[Negotiation] Round {current_round+1} bid: {result['offer']:.4f} USDC — {result['reasoning']}")

    # ── Seller side ────────────────────────────────────────────────

    def run_seller_loop(self, seller_address: str, floor_price_usdc: float,
                        service_description: str, on_deal: callable) -> None:
        handled_sessions = set()
        last_submitted = {}

        while True:
            time.sleep(self.poll_interval)
            try:
                open_sessions = self.client.get_open_sessions_for_seller(seller_address)
            except Exception as e:
                print(f"[Negotiation] Poll error: {e}")
                continue

            for session_id in open_sessions:
                session_hex = session_id.hex() if hasattr(session_id, 'hex') else bytes(session_id).hex()
                if session_hex in handled_sessions:
                    continue

                session = self.client.get_session(session_id)
                buyer_rep = self.client.get_reputation_display(session["buyer"])

                if buyer_rep < 1.0:
                    print(f"[Negotiation] Ignoring low-rep buyer: {buyer_rep:.1f}/10")
                    handled_sessions.add(session_hex)
                    continue

                print(f"[Negotiation] New session from buyer rep {buyer_rep:.1f}/10")
                handled_sessions.add(session_hex)

            for session_hex in list(handled_sessions):
                session_id_bytes = bytes.fromhex(session_hex)
                session = self.client.get_session(session_id_bytes)

                if session["status"] == "AGREED":
                    on_deal(session)
                    continue

                if session["status"] in ("EXPIRED", "CANCELLED"):
                    continue

                buyer_offers = session["buyerOffers"]
                seller_offers = session["sellerOffers"]
                current_round = session["currentRound"]
                round_key = f"{session_hex}:{current_round}"

                if (len(buyer_offers) > len(seller_offers) and
                    current_round == len(seller_offers) and
                    round_key not in last_submitted):

                    latest_bid = self._wei_to_usdc(buyer_offers[-1]) if buyer_offers else 0
                    bids_usdc = [self._wei_to_usdc(b) for b in buyer_offers]
                    asks_usdc = [self._wei_to_usdc(a) for a in seller_offers]
                    buyer_rep = self.client.get_reputation_display(session["buyer"])

                    if not buyer_offers:
                        initial_ask = floor_price_usdc * 2.5
                        ask_wei = self._usdc_to_wei(initial_ask)
                        self.client.submit_seller_ask(session_id_bytes, ask_wei)
                        last_submitted[round_key] = True
                        print(f"[Negotiation] Initial ask: {initial_ask:.4f} USDC")
                        continue

                    result = self.strategy.seller_strategy(
                        current_bid_usdc=latest_bid, floor_price_usdc=floor_price_usdc,
                        current_round=current_round, max_rounds=self.max_rounds,
                        buyer_reputation=buyer_rep, ask_history=asks_usdc,
                        bid_history=bids_usdc, service_description=service_description,
                    )

                    ask_wei = self._usdc_to_wei(result["offer"])
                    self.client.submit_seller_ask(session_id_bytes, ask_wei)
                    last_submitted[round_key] = True
                    print(f"[Negotiation] Round {current_round+1} ask: {result['offer']:.4f} USDC — {result['reasoning']}")

    # ── Helpers ────────────────────────────────────────────────────

    @staticmethod
    def _usdc_to_wei(usdc: float) -> int:
        return int(Web3.to_wei(usdc / 1000, "ether"))

    @staticmethod
    def _wei_to_usdc(wei: int) -> float:
        return float(Web3.from_wei(wei, "ether")) * 1000
